### ISSUE TITLE
Igbinary Session Support in PHP 7

### ISSUE THREAD

**Issue by jjeau75** [igbinary/igbinary#116]
We're running PHP 7.0.14 on FreeBSD and it appears the current igbinary for php7 doesn't include session support - as per below:

===============
igbinary support => enabled
igbinary version => 2.0.1
igbinary APC serializer ABI => no
igbinary session support => no
igbinary.compact_strings => On => On
===============

Has anyone successfully got it installed and working with sessions for PHP7?

We've tried everything we can think of and still no joy. The output from a "make test" gives:

TEST RESULT SUMMARY
---------------------------------------------------------------------
Exts skipped    :    0
Exts tested     :    9
---------------------------------------------------------------------

Number of tests :   68                59
Tests skipped   :    9 ( 13.2%) --------
Tests warned    :    0 (  0.0%) (  0.0%)
Tests failed    :    0 (  0.0%) (  0.0%)
Expected fail   :    0 (  0.0%) (  0.0%)
Tests passed    :   59 ( 86.8%) (100.0%)
---------------------------------------------------------------------
Time taken      :    1 seconds

With the main failure being:

> SKIP Check for serialization handler [tests/igbinary_015.phpt] reason: session extension not loaded
> SKIP Check for serialization handler, ini-directive [tests/igbinary_015b.phpt] reason: session extension not loaded

Any help would be greatly appreciated!

Thanks.

**Comment by TysonAndre** [comment]
igbinary has full php7 support. This includes support for serializing sessions in both php5 and php7.

1. What is the output of `php-config`? The --configure-options should contain --enable-session
2. Did you install the php-session package (Or whatever it is on your distro)
3. Add `extension=session.so` if it isn't in your /etc/php.ini file (or whatever is returned by https://secure.php.net/manual/en/function.php-ini-loaded-file.php)

**Comment by jjeau75** [comment]
Thanks for the quick reply.

Yes - we have the php-session package and that is working fine.

Output of php-config doesn't include --enable-session - and that doesn't surprise me as we are using the session package so I would not have expected php to have been compiled with --enable-session.

Are you suggesting we need to recompile php?

When we try ./configure --enable-session for the igbinary package - we get:
> configure: WARNING: unrecognized options: --enable-session

Thanks.

**Comment by jjeau75** [comment]
Note also our previous php installs (php 5.6) which we also used igbinary with did not have --enable-session as p[art of their compile.

**Comment by TysonAndre** [comment]
Yes, you need to compile the session module statically into php - shared won't work.

Built in session serializers (wddx) likely work the same way.

The change was made to fix a different issue in https://github.com/igbinary/igbinary/issues/77 (igbinary tests would fail if the igbinary session module was dynamically loadable, but the session module wasn't loaded at the time it was run).

**Comment by djGrrr** [comment]
Why would the session extension need to be statically compiled for igbinary to add it's serialization handler? This has always worked fine, and the wddx extension does it when session is loaded as shared extension just fine. The session module just needs to be loaded first, which has always been the case. 

The wddx module get's the same loading error as in #77 if the session extension is not loaded first

```
PHP Warning:  PHP Startup: Unable to load dynamic library '/usr/local/lib/php/20151012/wddx.so' - /usr/local/lib/php/20151012/wddx.so: Undefined symbol "ps_globals" in Unknown on line 0

Warning: PHP Startup: Unable to load dynamic library '/usr/local/lib/php/20151012/wddx.so' - /usr/local/lib/php/20151012/wddx.so: Undefined symbol "ps_globals" in Unknown on line 0
```

But if you make sure session is one of the first modules loaded:
```
Session Support => enabled
Registered save handlers => files user memcached redis rediscluster redis rediscluster
Registered serializer handlers => php_serialize php php_binary wddx

```
It would seem that the whole removing dynamic session support from igbinary was based on false information.

**Comment by TysonAndre** [comment]
> Why would the session extension need to be statically compiled for igbinary to add it's serialization handler? (@djGrrr)

igbinary can be used for various other things, such as (1) memcached serialization, (2) apcu serialization, and (3) redis serialization. Each of those things can be installed and used without installing/using the session extension.

---

Possibilities:

1. Try to detect if the session module is loaded already, and conditionally use the related functions - Probably not portable across operating systems
2. Work with PHP maintainers to try to establish a cross-platform way to register a session serializer, in a way which does not require the session module to be loaded/statically compiled into php. At earliest, this may be released in 7.2 a year from now.
   (If session module was not loaded, it would be a no-op)
3. Split up igbinary into 2 extensions, igbinary and igbinary_session. Installing igbinary_session through a package manager(deb, etc.) would also install the dependency of phpXYZ-igbinary and phpXYZ-session.
   igbinary_session would unconditionally register igbinary as a session serializer.
   (I don't really like this, and it would make upgrading existing installations of igbinary or sessions difficult/unintuitive)

**Comment by TysonAndre** [comment]
Any thoughts from the other igbinary maintainers?

**Comment by mqudsi** [comment]
Hello,

I modified the Makefile for `make test` to add `extension=/path/to/session.so` to try and see what happens when the session module is dynamically loaded when it isn't statically built in to PHP (running PHP 7.1.8).

Instead of being skipped because the `session` module isn't loaded, test 15 now fails with the following:

```
 Warning: ini_set(): Cannot find serialization handler 'igbinary' in /usr/home/mqudsi/php-ext/igbinary/tests/igbinary_015.php on line 32
```

The previous tests pass OK and igbinary is configured with `./configure --enable-igbinary`

Any clues?

**Comment by TysonAndre** [comment]
Aside: https://bugs.php.net/bug.php?id=34746 is a bug report from 2005 with a similar issue. sqlite owners declared it as 'not a bug'. ext/wddx does the same thing.

Once again, https://github.com/igbinary/igbinary/issues/116#issuecomment-271786157

As for mqudsi, no idea, and not enough information to reproduce (PHP version, igbinary version, etc. missing, not sure of steps used to build php, or what the steps used to build igbinary were). Statically loading is recommended. 

- Was the .so file compiled with a different PHP installation than the one used?

**Users can manually work around this in the following way**:

- If you need to dynamically load the session module, and use igbinary, replace `HAVE_DL_SESSION && !defined(COMPILE_DL_SESSION)` with `HAVE_DL_SESSION` in src/php5/igbinary.c and src/php7/igbinary.c 
- The COMPILE_DL_SESSION check was added because php would crash if session was not loaded or was not loaded first.

  If you have session as a module (i.e. not static), make sure that the line `extension=session.so` exists if igbinary is going to be loaded, and it occurs *above* `extension=igbinary.so` (Or the dll equivalent), so that igbinary's dependency on the session module is met.

**Comment by TysonAndre** [comment]
> ./configure --enable-session for the igbinary package

`--enable-session` is a flag for compiling the php-src folder


> Why would the session extension need to be statically compiled for igbinary to add it's serialization handler? This has always worked fine, and the wddx extension does it when session is loaded as shared extension just fine.

Why: The check was changed in a recent igbinary release to fix a crash, and behave more like built in php session modules. This can be manually undone with steps described in https://github.com/igbinary/igbinary/issues/116#issuecomment-322012617

The comment doesn't mention the OS or PHP version involved for wddx, or who packaged wddx(and what steps were used), or how it was validated wddx was still used for data serialization, so this doesn't help with fixing it. wddx has the same check to reject if if session is a module https://github.com/php/php-src/blob/PHP-7.1.8/ext/wddx/wddx.c#L263

- The fact that ext/wddx was in the php-src folder, unlike igbinary, may be related to any differences in behavior, but probably not.

**Comment by mqudsi** [comment]
@TysonAndre thanks for chiming in! I'm running PHP 7.1.8 from the FreeBSD ports, which is _not_compiled with php-session statically linked in. 

Their configuration:
```
'./configure' '--with-layout=GNU' '--localstatedir=/var' '--with-config-file-scan-dir=/usr/local/etc/php' '--disable-all' '--enable-libxml' '--enable-mysqlnd' '--with-libxml-dir=/usr/local' '--with-pcre-regex=/usr/local' '--program-prefix=' '--enable-fpm' '--with-fpm-user=www' '--with-fpm-group=www' '--enable-dtrace' '--prefix=/usr/local' '--mandir=/usr/local/man' '--infodir=/usr/local/info/' '--build=amd64-portbld-freebsd11.0' 'build_alias=amd64-portbld-freebsd11.0' 'CFLAGS=-O2 -pipe -fstack-protector -fno-strict-aliasing' 'CPPFLAGS=' 'CPP=cpp
```

I am building igbinary from `master`, and in my actual php configuration, I have session.so loaded before igbinary.so (`session.so.ini` is priority 18, `igbinary.so.ini` is priority 30, but also tried with a simple PHP.ini). If I include session.so manually prior to including igbinary.so I get an error/warning about session.so already being included. So that should all check out.

**Comment by mqudsi** [comment]
@TysonAndre do you mean replace `HAVE_DL_SESSION && !defined(COMPILE_DL_SESSION)` with `HAVE_DL_SESSION`?

I replaced it with both `HAVE_PHP_SESSION` and `HAVE_DL_SESSION`, but in both cases that did not take care of the skipped test.

**Comment by TysonAndre** [comment]
Now that I think about it, `extension_loaded` may work for unix and linux, where symbols are lazily loaded (I think, not familiar with how every possible shared library implementation)
- i.e. referencing an unavailable symbol would only abort where the symbol is accessed, not when igbinary.so is loaded

There might be issues on Windows if the session module is an unavailable DLL. I'd have to check.

If `zend_hash_exists(&module_registry, (zend_string for "session")` is false, then the session module isn't loaded dynamically or statically

```php
/* {{{ proto bool extension_loaded(string extension_name)
   Returns true if the named extension is loaded */
ZEND_FUNCTION(extension_loaded)
{
    zend_string *extension_name;
    zend_string *lcname;

    if (zend_parse_parameters(ZEND_NUM_ARGS(), "S", &extension_name) == FAILURE) {
        return;
    }

    lcname = zend_string_tolower(extension_name);
    if (zend_hash_exists(&module_registry, lcname)) {
        RETVAL_TRUE;
    } else {
        RETVAL_FALSE;
    }
    zend_string_release(lcname);
}
/* }}} */
```

---

So this would be changed to check for session being statically loaded, or for that module being found at runtime.

```c
#if HAVE_PHP_SESSION && !defined(COMPILE_DL_SESSION)
    php_session_register_serializer("igbinary",
        PS_SERIALIZER_ENCODE_NAME(igbinary),
        PS_SERIALIZER_DECODE_NAME(igbinary));
#endif
```

**Comment by andypost** [comment]
As I got it the APCu support also removed from 3.0 #153 so both looks like needs work
```
igbinary support => enabled
igbinary version => 3.2.2RC1
igbinary APCu serializer ABI => 0
igbinary session support => no
```

**Comment by TysonAndre** [comment]
No, https://pecl.php.net/package/APCu is a replacement for https://pecl.php.net/package/APC and they're different extensions. igbinary supports APCu.

In order to support sessions, session support must be statically compiled. I'd looked into other solutions but there was no interest in modifying php-src itself to support this - this is what all session handlers do.

```c
/**
 * This provides a serializer encode function for PHP's session module (using igbinary),
 * if igbinary was compiled with session support.
 *
 * Session support has to be statically compiled into php to use igbinary,
 * due to the lack of a cross-platform way to register a session serializer/unserializer
 * when the session module isn't available.
 */
PS_SERIALIZER_ENCODE_FUNC(igbinary)
```

**Comment by rapsys** [comment]
As told in bug report #349 it's sad, it introduce a regression for linux user for an issue concerning only windows dll and make test if I understand well.

I patched mageia distribution backport and cauldron package to workaround this feature drop that was used in production.

Suggesting to build all modules as static is not wise for production and may be a problem for security.

It prevent the operator to drop impacted php module until a patch and/or full rebuild is done...

**Comment by TysonAndre** [comment]
> it introduce a regression for linux user for an issue concerning only windows dll and make test if I understand well.

Using an undefined symbol is undefined behavior in C as far as I'm aware. Windows is the most prominent example of using an undefined symbol failing, but I don't have the information needed to say it's the only example. Even if it works in the linux distributions tested, I wasn't confident it'd work in other unixes (FreeBSD, etc.), or in future releases of those linux distributions.

**Comment by rapsys** [comment]
Can't you consider to allow it to be enabled by a configure option ?

Like --enable-shared or something ?
